Improved HazelcastStarter for SplitBrainTestSupport and Node access

Donnerbart opened the pull request:
* Added new constructor classes for `HazelcastStarter`
* Added new answer classes for `HazelcastStarter`
* Improved `ReflectionUtils` for `HazelcastStarter`
* Added download of test JARs to `HazelcastVersionLocator`
* Improved `HazelcastStarter` for `SplitBrainTestSupport` and `Node` access
* Fixed `HazelcastInstanceImpl` casting in cache tests

Donnerbart commented:
run-lab-run

vbekiaris commented on hazelcast/src/test/java/com/hazelcast/test/starter/constructor/AbstractConfigConstructor.java:
```diff
@@ -105,6 +107,12 @@ static Object cloneConfig(Object thisConfigObject, ClassLoader classloader) thro
+                            "com.hazelcast.core.RingbufferStore");
+                } else if (returnType.getName().startsWith("com.hazelcast.core.QueueStore")) {
+                    cloneStoreInstance(classloader, method, setter, thisConfigObject, otherConfigObject,
+                            "com.hazelcast.core.QueueStore");
```
similarly to comment above, this one will also match `com.hazelcast.core.QueueStoreFactory`

Donnerbart replied:
I've added a `QueueStoreCompatibilityTest` on the EE side (with a little tweak to really test all Hazelcast versions):
```java
@RunWith(EnterpriseSerialJUnitClassRunner.class)
@Category(CompatibilityTest.class)
public class QueueStoreCompatibilityTest extends QueueStoreTest {

    @Override
    protected HazelcastInstance createHazelcastInstance(Config config) {
        // we create the factory with 3 instances, to get an Hazelcast instance with the CURRENT_VERSION as well
        // and return this to the test (instance[0] and instance[1] are proxied instances)
        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(3);
        HazelcastInstance[] instances = factory.newInstances(config);
        warmUpPartitions(instances);
        return instances[2];
    }
}
```
This worked well, I just made the `AbstractConfigConstructor.cloneConfig()` method more concrete about the supported types:
```java
if (returnTypeName.equals("com.hazelcast.core.RingbufferStore")
        || returnTypeName.equals("com.hazelcast.core.RingbufferStoreFactory")
        || returnTypeName.equals("com.hazelcast.core.QueueStore")
        || returnTypeName.equals("com.hazelcast.core.QueueStoreFactory")) {
    cloneStoreInstance(classloader, method, setter, thisConfigObject, otherConfigObject, returnTypeName);
}
```

There was one new error, originated in the compatibility framework: `ClassCastException: [B cannot be cast to [Ljava.lang.Object`, thrown in the `HazelcastProxyFactory.proxyObjectForStarter()` method. It was originated in the `if (arg.getClass().isArray())` code, which I pulled out in a new method. It needed a fix to support primitive arrays (I've added `byte[]`, `int[]` and `long[]` for now):
```java
private static Object toArray(ClassLoader targetClassLoader, Object arg) throws ClassNotFoundException {
    if (arg instanceof byte[]) {
        byte[] srcArray = ((byte[]) arg);
        byte[] targetArray = new byte[srcArray.length];
        arraycopy(srcArray, 0, targetArray, 0, srcArray.length);
        return targetArray;
    } else if (arg instanceof int[]) {
        int[] srcArray = ((int[]) arg);
        int[] targetArray = new int[srcArray.length];
        arraycopy(srcArray, 0, targetArray, 0, srcArray.length);
        return targetArray;
    } else if (arg instanceof long[]) {
        long[] srcArray = ((long[]) arg);
        long[] targetArray = new long[srcArray.length];
        arraycopy(srcArray, 0, targetArray, 0, srcArray.length);
        return targetArray;
    }
    Object[] srcArray = ((Object[]) arg);
    Class<?> targetClass = targetClassLoader.loadClass(srcArray.getClass().getComponentType().getName());
    Object[] targetArray = (Object[]) Array.newInstance(targetClass, srcArray.length);
    for (int i = 0; i < srcArray.length; i++) {
        targetArray[i] = proxyObjectForStarter(targetClassLoader, srcArray[i]);
    }
    return targetArray;
}
```

The `QueueStoreTest.testIssue1401QueueStoreWithTxnPoll()` fails most of the time, but it's the only one using a `TransactionalQueue`, so I guess we're still missing some kind of support for it. I've ignored it in the `QueueStoreCompatibilityTest` for now.

Donnerbart commented:
> hm, I reviewed the constructors commit but the PR changed? posting my comments so far.
Yes, sorry, I made changes to the answer package on Thursday. I've simplified the `createMapMock()` collection methods to not call `invoke()` recursively. And moved all the "return type based handling" after the `Object result = invokeDelegateMethod(delegateMethod, arguments);` line. That made everything much easier to read. All other packages were untouched by the last push.

I'll check your comments right now, e.g. I'll try to run the `QueueStoreTest` as compatibility test to check for the `QueueStoreFactory` support.

Donnerbart commented on hazelcast/src/test/java/com/hazelcast/test/starter/constructor/AbstractConfigConstructor.java:
```diff
@@ -105,6 +107,12 @@ static Object cloneConfig(Object thisConfigObject, ClassLoader classloader) thro
                 } else if (returnType.getName().startsWith("java") || returnType.isPrimitive()) {
                     Object thisSubConfigObject = method.invoke(thisConfigObject, null);
                     updateConfig(setter, otherConfigObject, thisSubConfigObject);
+                } else if (returnType.getName().startsWith("com.hazelcast.core.RingbufferStore")) {
```
I've added a `RingbufferStoreCompatibilityTest` on the EE side (with a little tweak to really test all Hazelcast versions):
```java
@RunWith(EnterpriseSerialJUnitClassRunner.class)
@Category(CompatibilityTest.class)
public class RingbufferStoreCompatibilityTest extends RingbufferStoreTest {

    @Override
    protected HazelcastInstance createHazelcastInstance(Config config) {
        // we create the factory with 3 instances, to get an Hazelcast instance with the CURRENT_VERSION as well
        // and return this to the test (instance[0] and instance[1] are proxied instances)
        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(3);
        HazelcastInstance[] instances = factory.newInstances(config);
        warmUpPartitions(instances);
        return instances[2];
    }
}
```
This worked well, I just made the `AbstractConfigConstructor.cloneConfig()` method more concrete about the supported types:
```java
if (returnTypeName.equals("com.hazelcast.core.RingbufferStore")
        || returnTypeName.equals("com.hazelcast.core.RingbufferStoreFactory")
        || returnTypeName.equals("com.hazelcast.core.QueueStore")
        || returnTypeName.equals("com.hazelcast.core.QueueStoreFactory")) {
    cloneStoreInstance(classloader, method, setter, thisConfigObject, otherConfigObject, returnTypeName);
}
```
There seem to be no spurious failures in the `RingbufferStoreCompatibilityTest`.

Donnerbart commented on hazelcast/src/test/java/com/hazelcast/test/starter/constructor/MemberImplConstructor.java:
```diff
@@ -0,0 +1,110 @@
+            //                                            String uuid, Map<String, Object> attributes, boolean liteMember,
+            //                                            int memberListJoinVersion, HazelcastInstanceImpl instance)
+            constructor = targetClass.getDeclaredConstructor(addressClass, memberVersionClass, Boolean.TYPE, String.class,
+                    Map.class, Boolean.TYPE); //, Integer.TYPE, hazelcastInstanceImplClass);
```
Good catch, I've deleted the dead code. I think in the future it would be good to support the constructor with the `memberListJoinVersion`, so we don't lose this information. But also `MemberImpl` could need a little cleanup, since it uses `instance.node.address` and `instance.node.nodeEngine`, which is a bad style for compatibility test involved code.

Donnerbart commented:
run-lab-run
